system/lib/pthread/proxying.c

```diff
@@ -330,24 +335,26 @@ void emscripten_proxy_execute_queue(em_proxying_queue* q) {
 
   pthread_mutex_lock(&q->mutex);
   task_queue* tasks = get_tasks_for_thread(q, pthread_self());
-  if (tasks == NULL || tasks->processing) {
-    // No tasks for this thread or they are already being processed.
-    goto end;
-  }
-  // Found the task queue; process the tasks.
-  tasks->processing = 1;
-  while (!task_queue_is_empty(tasks)) {
-    task t = task_queue_dequeue(tasks);
-    // Unlock while the task is running to allow more work to be queued in
-    // parallel.
-    pthread_mutex_unlock(&q->mutex);
-    t.func(t.arg);
-    pthread_mutex_lock(&q->mutex);
+  pthread_mutex_unlock(&q->mutex);
+
+  if (tasks != NULL) {
+    pthread_mutex_lock(&tasks->mutex);
+    if (!tasks->processing) {
+      // Found the task queue; process the tasks.
```

Comment:
```suggestion
      // Found the task queue, and it is not already being processed; process the tasks.
```
(this moves some of the text of the removed comment above to here)